components/supervisor/pkg/terminal/terminal.go

```diff
@@ -146,43 +147,56 @@ func (m *Mux) doClose(alias string, gracePeriod time.Duration) error {
 	return nil
 }
 
-func (term *Term) gracefullyShutdownProcess(gracePeriod time.Duration) error {
+func (term *Term) gracefullyShutdownProcess(wg *sync.WaitGroup) error {
 	if term.Command.Process == nil {
-		// process is alrady gone
+		// process is already gone
 		return nil
 	}
-	if gracePeriod == 0 {
-		return term.shutdownProcessImmediately()
-	}
-
-	err := term.Command.Process.Signal(unix.SIGTERM)
-	if err != nil {
-		return err
+	pid := term.Command.Process.Pid
+	var subwg sync.WaitGroup
+	processes, err := procfs.AllProcs()
+	if err == nil {
+		// terminate direct child processes
+		for _, p := range processes {
```

Comment:
I'm not sure it is a way to go, the problem here is if someone has a process which start another process and they have a contract between each other based on signals. We will break it because we can terminate child process before parent.

The only real direct child process here is bash.

Reply:
You mean it is not ok to assume that the main terminal process is a shell?

Reply:
> You mean it is not ok to assume that the main terminal process is a shell? 

In case of regular workspace tasks that is safe to assume.
AFAIK supervisor executes `dockerd` in a terminal but not in a shell.